src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,312 @@
+
+-- cf.`add_monoid_hom.add_comm_monoid`
+instance : add_comm_monoid (centroid_hom α) :=
+coe_to_add_monoid_hom_injective.add_comm_monoid _ to_End_zero to_End_add to_End_nsmul
+
+instance : has_nat_cast (centroid_hom α) :=
+{ nat_cast := λ n, n • 1 }
+
+lemma nat_cast_apply (n : ℕ) (m : α):
+  (n : centroid_hom α) m = n • m := rfl
+
+@[simp] lemma coe_nat_cast (n : ℕ) : ⇑(n : centroid_hom α) = n • id := rfl
+
+@[simp] lemma to_End_one : (1 : centroid_hom α).to_End = 1 := rfl
+@[simp] lemma to_End_mul (x y : centroid_hom α) : (x * y).to_End = x.to_End * y.to_End := rfl
+@[simp] lemma to_End_pow (x : centroid_hom α) (n : ℕ) : (x ^ n).to_End = x.to_End ^ n :=
+by { ext, refl }
+@[simp, norm_cast] lemma to_End_nat_cast (n : ℕ) : (n : centroid_hom α).to_End = ↑n := rfl
+
+-- cf `add_monoid.End.semiring`
+instance : semiring (centroid_hom α) :=
+to_End_injective.semiring _ to_End_zero to_End_one to_End_add to_End_mul
+  to_End_nsmul to_End_pow to_End_nat_cast
+
+lemma comp_mul_comm (T S : centroid_hom α) (a b : α) : (T ∘ S) (a * b) = (S ∘ T) (a * b) :=
+by rw [comp_app, map_mul_right, map_mul_left, ←map_mul_right, ←map_mul_left]
+
+end non_unital_non_assoc_semiring
+
+section non_unital_non_assoc_ring
+variables [non_unital_non_assoc_ring α]
+
+/-- Negation of `centroid_hom`s as a `centroid_hom`. -/
+instance : has_neg (centroid_hom α) :=
+⟨λ f,
+  { map_mul_left' := by simp [map_mul_left],
+    map_mul_right' := by simp [map_mul_right],
+    .. (-f : α →+ α) }⟩
+instance : has_sub (centroid_hom α) :=
+⟨λ f g,
+{ map_mul_left' := λ a b, by simp [map_mul_left, mul_sub],
+  map_mul_right' := λ a b, by simp [map_mul_right, sub_mul],
+  .. (f - g : α →+ α) }⟩
+
+instance has_zsmul : has_smul ℤ (centroid_hom α) :=
+⟨λ n f,
+  { map_mul_left' := λ a b,
+      by { change n • f (a * b) = a * n • f b, rw [map_mul_left f, ←mul_smul_comm] },
+    map_mul_right' := λ a b,
+      by { change n • f (a * b) = n • f a * b, rw [map_mul_right f, ←smul_mul_assoc] },
+    .. (n • f : α →+ α) }⟩
+
+instance : has_int_cast (centroid_hom α) :=
+{ int_cast := λ z, z • 1 }
+
+lemma int_cast_apply (z : ℤ) (m : α) :
+  (z : centroid_hom α) m = z • m := rfl
+
+@[simp] lemma to_End_neg (x : centroid_hom α) : (-x).to_End = -x.to_End := rfl
+@[simp] lemma to_End_sub (x y : centroid_hom α) : (x - y).to_End = x.to_End - y.to_End := rfl
+lemma to_End_zsmul (x : centroid_hom α) (n : ℤ) : (n • x).to_End = n • x.to_End := rfl
+
+instance : add_comm_group (centroid_hom α) :=
+to_End_injective.add_comm_group _ to_End_zero to_End_add to_End_neg to_End_sub
+  to_End_nsmul to_End_zsmul
+
+@[simp] lemma coe_int_cast (z : ℤ) : ⇑(z : centroid_hom α) = z • id := rfl
+@[simp] lemma coe_neg (f : centroid_hom α) : ⇑(-f) = -f := rfl
+@[simp] lemma coe_sub (f g : centroid_hom α) : ⇑(f - g) = f - g := rfl
+
+@[simp] lemma neg_apply (f : centroid_hom α) (a : α) : (-f) a = - f a := rfl
+@[simp] lemma sub_apply (f g : centroid_hom α) (a : α) : (f - g) a = f a - g a := rfl
```

Comment:
Could you try keeping the `coe_whatever` lemma close to the `whatever_apply` one? That is, either group `coe_int_cast` and `int_cast_apply` together after `has_int_cast (centroid_hom α)`  or move `int_cast_apply` just before `neg_apply`.

Reply:
Done.